### PR TITLE
fix: update cached fromStore value when the package has been downloaded

### PR DESCRIPTION
**Pull request by etamponi**
Also added a test for it.

### CODE REVIEW

File: src/packageRequester.ts [hunk 1]
```diff
@@ -376,6 +376,27 @@ function fetchToStore (
       })
     }
 
+    // when fetchingFiles resolves, the cached result has to set fromStore to true, without
+    // affecting previous invocations: so we need to replace the cache.
+    fetchingFiles.promise.then((res) => {
+      const tmp = ctx.fetchingLocker.get(opts.pkgId) as {
```
Comment (zkochan) [inline src/packageRequester.ts]:
```typescript
if (res.fromStore) return
```

File: src/packageRequester.ts [hunk 2]
```diff
@@ -376,6 +376,27 @@ function fetchToStore (
       })
     }
 
+    // when fetchingFiles resolves, the cached result has to set fromStore to true, without
+    // affecting previous invocations: so we need to replace the cache.
+    fetchingFiles.promise.then((res) => {
```
Comment (zkochan) [inline src/packageRequester.ts]: I don't rem

Comment (zkochan) [inline src/packageRequester.ts]: you can use destructuring here, so

```typescript
fetchingFiles.promise.then(({filenames}) => {
```

File: src/packageRequester.ts (hunk 1 above)
Comment (etamponi) [reply src/packageRequester.ts]: right!

File: src/packageRequester.ts [hunk 3]
```diff
@@ -376,6 +376,27 @@ function fetchToStore (
...
+        inStoreLocation: string,
+      }
+
+      ctx.fetchingLocker.set(opts.pkgId, {
+        fetchingFiles: Promise.resolve({
+          filenames: res.filenames,
+          fromStore: true,
+        }),
+        fetchingRawManifest: tmp.fetchingRawManifest && removeKeyOnFail(tmp.fetchingRawManifest),
+        finishing: removeKeyOnFail(tmp.finishing),
```
Comment (zkochan) [inline src/packageRequester.ts]: there is no need to wrap these fields with removeKeyOnFail again. Just do

```typescript
fetchingRawManifest: tmp.fetchingRawManifest,
finishing: tmp.finishing,
```

**Comment by zkochan** [comment]
LGTM, after merging I'll add more comments to highlight that this hack is only needed for correct reporting of `pnpm server`

**Pull request merged**